File: taurus-mobile/android/app/src/main/java/com/numenta/taurus/data/Tweet.java
/*
 * Numenta Platform for Intelligent Computing (NuPIC)
 * Copyright (C) 2015, Numenta, Inc.  Unless you have purchased from
 * Numenta, Inc. a separate commercial license for this software code, the
 * following terms and conditions apply:
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses.
 *
 * http://numenta.org/licenses/
 *
 */

package com.numenta.taurus.data;

import com.numenta.core.utils.DataUtils;

import java.io.Serializable;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Helper class representing one "tweet" record
 */
public class Tweet implements Serializable {

    private String _id;

    private long _created;

    private long _aggregated;

    private String _userId;

    private String _userName;

    private String _text;

    // Regex used to filter duplicate tweets. The current logic removes optional "RT " from the
    // beginning of the text and optional URL from the end of the text.
    //
    // For example:
    //   RT @Company: Watch it again:  the story of John Doe.\nhttps://t.co/yWd0iW3GdJ
    //   RT @Company: Watch it again:  the story of John Doe.\nhttps://t.co/sdSADsSDfs
    //   @Company: Watch it again:  the story of John Doe.\nhttps://t.co/yWd0iW3GdJ
    //   @Company: Watch it again:  the story of John Doe.\nhttps://t.co/yWd0Ed0WAC
    //
    // All translate to:
    //   @Company: Watch it again:  the story of John Doe.
    //
    private static final Pattern TWEET_TEXT_FILTER = Pattern.compile("(?:^RT\\s+)?(.+?)(?:\\s+https?:\\/\\/\\S+\\s*)*$");

    // FIXME: Optimize storage for these Runtime values.
    private int _aggregatedCount;

    private int _retweetCount;

    private int _retweetTotal;

    public Tweet(String id, long aggregated, long created, String userId, String userName,
            String text, int retweetTotal) {
        _text = text;
        _userName = userName;
        _userId = userId;
        _created = created;
        // Make sure aggregated values are rounded to closest 5 minute interval
        _aggregated = DataUtils.floorTo5minutes(aggregated);
        _id = id;
        _retweetTotal = retweetTotal;
    }

    /**
     * Return twitter text without retweets (RT) and other decorators
     */
    public String getCanonicalText() {
        Matcher result = TWEET_TEXT_FILTER.matcher(_text);
        if (result.matches() && result.groupCount() == 1) {
            return result.group(1);
        }
        return _text;
    }

    public String getId() {
        return _id;
    }

    public long getAggregated() {
        return _aggregated;
    }

    public long getCreated() {
        return _created;
    }

    public String getUserId() {
        return _userId;
    }

    public String getUserName() {
        return _userName;
    }

    public String getText() {
        return _text;
    }

    public void setText(String text) {
        _text = text;
    }

    public int getAggregatedCount() {
        return _aggregatedCount;
    }

    public void setAggregatedCount(int aggregatedCount) {
        _aggregatedCount = aggregatedCount;
    }

    public int getRetweetCount() {
        return _retweetCount;
    }

    public void setRetweetCount(int count) {
        _retweetCount = count;
    }

    public int getRetweetTotal() {
        return _retweetTotal;
    }
}

File: taurus-mobile/ios/src/data/Tweet.swift
// Numenta Platform for Intelligent Computing (NuPIC)
// Copyright (C) 2015, Numenta, Inc.  Unless you have purchased from
// Numenta, Inc. a separate commercial license for this software code, the
// following terms and conditions apply:
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero Public License version 3 as
// published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Affero Public License for more details.
//
// You should have received a copy of the GNU Affero Public License
// along with this program.  If not, see http://www.gnu.org/licenses.
//
// http://numenta.org/licenses/


/**
* Represent one tweet
*/
class Tweet  {
    var id: String!
    var created: Int64
    var aggregated: Int64
    var userId: String!
    var userName: String!
    var text: String!
    var hasLinks: Bool
    var cannonicalText: String!
  
    var aggregatedCount: Int32
    var retweetCount: Int32
    var retweetTotal: Int32
    
    
    //  // Match "RT @tags" from the beginning of the text.
    let RT_REGEX: String = "^\\s*RT\\s+(@[a-zA-Z][_a-zA-Z0-9]*\\s*)*"
    //  // From left side – Match @, #, $ up to colon symbol
    let LEFT_HASHTAG_UP_TO_COLON_REGEX: String! = "^\\s*([@#$][a-zA-Z][_a-zA-Z0-9]*\\s?)*:\\s*"
    //  // From left side – Match @, #, $ up to last
    let LEFT_HASHTAG_UP_TO_LAST_REGEX : String =  "^\\s*([@#$][a-zA-Z][_a-zA-Z0-9]*\\s*){2,}"
    //  // From right side – Match @, #, $ when followed by letter, not a number
    let RIGHT_HASHTAG_REGEX: String = "\\s*([@#$][a-zA-Z][_a-zA-Z0-9]*\\s*)+\\s*$"
    //  // Match "http" or "https" URLs
    let LINKS_REGEX: String = "\\s*https?:\\/\\/\\S+\\s*"
    //  // Match "..." at the end of the text
    let DOT_DOT_DOT_REGEX: String = "\\s*\\.{2,}\\s*$"
    //  // Match 2 or more spaces
    let TWO_OR_MORE_SPACES_REGEX: String = "\\s+"
    

    init(id: String!, aggregated: Int64, created: Int64, userId: String!, userName: String!, text: String!, retweetTotal: Int32) {
        self.userName = userName
        self.userId = userId
        self.created = created
        //  // Make sure aggregated values are rounded to closest 5 minute interval
        self.aggregated = DataUtils.floorTo5minutes(aggregated)
        self.id = id
        self.retweetTotal = retweetTotal
        
        self.text = text;
        self.cannonicalText = text
        self.hasLinks = false
        self.aggregatedCount = 0
        self.retweetCount = 0
        
        self.cannonicalText = self.makeCannonicalText(text)

        
              //  // Clean up HTML encoded text
       var cleanedText = text.stringByReplacingOccurrencesOfString("&amp;", withString: "&")
        cleanedText = cleanedText.stringByReplacingOccurrencesOfString("&quot;", withString: "\"")
        cleanedText = cleanedText.stringByReplacingOccurrencesOfString("&lt;", withString: "<")
        cleanedText = cleanedText.stringByReplacingOccurrencesOfString("&gt;", withString: ">")
 
        self.text = cleanedText
        
        
    }

    /**
    * Return twitter text without retweets (RT) and other decorators. The current logic is:
    *
    * <ul>
    * <li> Remove "RT " from the beginning of the text up to the colon.</li>
    * <li> Remove all links from the text.</li>
    * <li> From left side – remove @, #, $ up to colon symbol.</li>
    * <li> If no colon – remove @, #, $ when there are multiple in a row (up to last one).</li>
    * <li>From right side – remove @, #, $ (only remove $ when followed by letter, not a number)
    * up to text</li>
    * </ul>
    * <p>
    * For example:
    * <ol>
    * <li>
    * <b>Original:</b> "RT @Kelly_Evans: Salesforce $CRM just reopened over at Post 6 and surging
    * 13% on BBG reports of possible suitors @NYSE"
    * <br>
    * <b>Condensed:</b> "Salesforce $CRM just reopened over at Post 6 and surging 13% on BBG
    * reports of possible suitors"
    * </li>
    * <li>
    * <b>Original:</b> "RT @hblodget: Someone offered to buy Salesforce! RT@SAI: Salesforce
    * has hired advisors to fend off takeover offers $CRM http://t.co/DGfSez"
    * <br>
    * <b>Condensed: </b>"Someone offered to buy Salesforce! RT@SAI: Salesforce has hired
    * advisors to fend off takeover offers $CRM <b>link</b>"
    * </li>
    * <li>
    * <b>Original:</b> "RT @BloombergDeals: Breaking on Bloomberg: http://t.co/nuMtaw43j is
    * said to hire bankers to field takeover interest. Terminal link: http://t.co/ssfSEca3rk"
    * <br>
    * <b>Condensed:</b> "Breaking on Bloomberg: is said to hire bankers to field takeover
    * interest. Terminal link: <b>link</b>"
    * </li>
    * <li>
    * <b>Original:</b> "RT @OptionHawk: $CRM takeover talks, whoa!"
    * <br>
    * <b>Condensed:</b> "$CRM takeover talks, whoa!"
    * </li>
    * <li>
    * <b>Original:</b> "$SUTI Has now seen 400% gains in the past 2 weeks!
    * Special update: http://t.co/DGfSez3QA $ESRX $COST $BIIB"
    * <br>
    * <b>Condensed:</b> "Has now seen 400% gains in the past 2 weeks! Special update: <b>link</b>"
    * </li>
    * <li>
    * <b>Original:</b> "MCD McDonalds Corp. Volume http://t.co/sfda3SDA $MCD $DG $MRO $FCEL
    * $MCD #stock #tradeideas"
    * <br>
    * <b>Condensed:</b> "MCD McDonalds Corp. Volume <b>link</b>"
    * </li>
    * <li>
    * <b>Original:</b> "RT @wbznewsradio: NEW: @CharterCom to buy @TWC for $53.33 billion in
    * cash-and-stock deal. Charter will also buy @BrighHouseNow for $10B+"
    * <br>
    * <b>Condensed:</b> "NEW: @CharterCom to buy @TWC for $53.33 billion in cash-and-stock
    * deal. Charter will also buy @BrighHouseNow for $10B+"
    * </li>
    * </ol>
    * </p>
    */
    func makeCannonicalText( rawTweet : String)->String{
         var rawText: String = rawTweet
        
        
        // Remove ...
        
        let dotdotdotRegEx = try! NSRegularExpression(pattern: DOT_DOT_DOT_REGEX,
            options: [.CaseInsensitive])
        rawText = dotdotdotRegEx.stringByReplacingMatchesInString(rawText, options:[],
            range: NSMakeRange(0, rawText.characters.count ), withTemplate: "")
        
        // Links
        let linkRegEx = try! NSRegularExpression(pattern: LINKS_REGEX,
            options: [.CaseInsensitive])
        
       
        if ( linkRegEx.firstMatchInString(rawText, options:[],
            range: NSMakeRange(0, rawText.characters.count ) ) != nil ){
                
                self.hasLinks = true
                
                rawText = linkRegEx.stringByReplacingMatchesInString(rawText, options:[],
                    range: NSMakeRange(0, rawText.characters.count ), withTemplate: " ")
                
        }
        
        
        let rtRegEx = try! NSRegularExpression(pattern: RT_REGEX,
            options: [.CaseInsensitive])
        rawText = rtRegEx.stringByReplacingMatchesInString(rawText, options:[],
            range: NSMakeRange(0, rawText.characters.count ), withTemplate: " ")
       
        let leftHashRegEx = try! NSRegularExpression(pattern: LEFT_HASHTAG_UP_TO_COLON_REGEX,
            options: [.CaseInsensitive])
       

        
        if ( leftHashRegEx.firstMatchInString(rawText, options:[],
            range: NSMakeRange(0, rawText.characters.count ) ) != nil ){
        
                rawText = leftHashRegEx.stringByReplacingMatchesInString(rawText, options:[],
                    range: NSMakeRange(0, rawText.characters.count ), withTemplate: " ")
                
        }else{
            rawText = leftHashRegEx.stringByReplacingMatchesInString(rawText, options:[],
                range: NSMakeRange(0, rawText.characters.count ), withTemplate: "$1")

        }
        
        // Remove hash tags from the right
        let rightHashRegEx = try! NSRegularExpression(pattern: RIGHT_HASHTAG_REGEX,
            options: [.CaseInsensitive])
        rawText = rightHashRegEx.stringByReplacingMatchesInString(rawText, options:[],
            range: NSMakeRange(0, rawText.characters.count ), withTemplate: " ")
        
        
        let twospaceRegEx = try! NSRegularExpression(pattern: TWO_OR_MORE_SPACES_REGEX,
            options: [.CaseInsensitive])
        rawText = twospaceRegEx.stringByReplacingMatchesInString(rawText, options:[],
            range: NSMakeRange(0, rawText.characters.count ), withTemplate: " ")

        rawText =  rawText.stringByReplacingOccurrencesOfString("\\n", withString:"")
        rawText = rawText.stringByReplacingOccurrencesOfString("\\r", withString:"")
        
        
        
     // Remove duplicate tweets using specific set of rules. See "getCannonicalText"for rules
        
        
        
        
       return rawText
    
    }



}
